File: internal/limiter/tokenbucket/token_bucket.lua
-- KEYS[1]: bucket key
-- ARGV[1]: capacity (max tokens)
-- ARGV[2]: refill_rate (tokens per second)
-- ARGV[3]: cost (tokens this request consumes)
-- ARGV[4]: now (current time in milliseconds)
-- Returns: 1 if allowed, 0 if rejected

local key = KEYS[1]
local capacity = tonumber(ARGV[1])
local refill_rate = tonumber(ARGV[2])
local cost = tonumber(ARGV[3])
local now = tonumber(ARGV[4])

local data = redis.call("HMGET", key, "tokens", "last_refill")
local tokens = tonumber(data[1])
local last_refill = tonumber(data[2])

if tokens == nil then
    tokens = capacity
    last_refill = now
end

local elapsed = (now - last_refill) / 1000.0
local new_tokens = math.min(capacity, tokens + elapsed * refill_rate)

local ttl = math.ceil(capacity / refill_rate) * 2

if new_tokens < cost then
    redis.call("HSET", key, "tokens", new_tokens, "last_refill", now)
    redis.call("EXPIRE", key, ttl)
    return 0
end

redis.call("HSET", key, "tokens", new_tokens - cost, "last_refill", now)
redis.call("EXPIRE", key, ttl)
return 1

File: internal/limiter/tokenbucket/tokenbucket.go
package tokenbucket

import (
	"context"
	_ "embed"
	"time"

	goredis "github.com/redis/go-redis/v9"
)

//go:embed token_bucket.lua
var luaScript string

type TokenBucket struct {
	client     *goredis.Client
	script     *goredis.Script
	capacity   int
	refillRate float64 // tokens per second
}

func New(client *goredis.Client, capacity int, refillRate float64) *TokenBucket {
	return &TokenBucket{
		client:     client,
		script:     goredis.NewScript(luaScript),
		capacity:   capacity,
		refillRate: refillRate,
	}
}

func (tb *TokenBucket) Allow(ctx context.Context, key string, cost int) (bool, error) {
	return tb.allow(ctx, key, cost, time.Now().UnixMilli())
}

func (tb *TokenBucket) allow(ctx context.Context, key string, cost int, nowMs int64) (bool, error) {
	result, err := tb.script.Run(ctx, tb.client, []string{key},
		tb.capacity,
		tb.refillRate,
		cost,
		nowMs,
	).Int()
	if err != nil {
		return false, err
	}
	return result == 1, nil
}
